File: lcof/of007/Solution.go
package Solution

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

//	递归求解
func buildTree(pre []int, in []int) *TreeNode {
	if len(pre) == 0 || len(in) == 0 {
		return nil
	}
	//	找到跟节点位置
	mid := search(in, pre[0])
	return &TreeNode{
		Val:   pre[0],
		Left:  buildTree(pre[1:mid+1], in[:mid+1]),
		Right: buildTree(pre[mid+1:], in[mid+1:]),
	}
}
func search(nodes []int, val int) int {
	for p, v := range nodes {
		if v == val {
			return p
		}
	}
	return -1
}

File: lcof/of007/Solution.py
from typing import List


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        def recur(root, left, right):
            if left > right: return  # 递归终止
            node = TreeNode(preorder[root])  # 建立根节点
            i = dic[preorder[root]]  # 划分根节点、左子树、右子树
            node.left = recur(root + 1, left, i - 1)  # 开启左子树递归
            node.right = recur(i - left + root + 1, i + 1, right)  # 开启右子树递归
            return node  # 回溯返回根节点

        dic, preorder = {}, preorder
        for i in range(len(inorder)):
            dic[inorder[i]] = i
        return recur(0, 0, len(inorder) - 1)


if __name__ == '__main__':
    Solution()
